crypto/err/err.c

```diff
@@ -744,6 +774,9 @@ int ERR_get_next_error_library(void)
     ret = int_err_library_number++;
     CRYPTO_THREAD_unlock(err_string_lock);
     return ret;
+#else
+    return int_err_library_number;
+#endif
```

Comment:
This change is not ok.  The error library number isn't just about the strings, it also affects the error code itself, and is used to distinguish where an error comes from (extractable with `ERR_LIB`)

Reply:
For example, every engine usually get themselves a unique error library number, to be able to report their own errors.  With this change, it's not possible which engine a particular engine generated error comes from, and since they can have overlapping reason codes, it's therefore not even possible to distinguish what the error was

Reply:
The same goes for providers, btw...  See this:

https://github.com/openssl/openssl/blob/927d0566ded0dff9d6c5abc8a40bb84068446b76/crypto/provider_core.c#L555

Reply:
> This change is not ok. The error library number isn't just about the strings, it also affects the error code itself, and is used to distinguish where an error comes from (extractable with `ERR_LIB`)

Ah, I already pondered this question in https://github.com/openssl/openssl/pull/17975#issuecomment-1082315309. But a (probably too) quick search in 1.1.1 didn't bring up any indication that the number was used other than looking up the string. And @bernd-edlinger agreed that it was ok. 

> As this PR currently stands, -1

No problem, it's out of question that this part of my change needs to be reverted (in both PRs).

Reply:
@levitte may I ask you that you add this as a formal review change request to #17975 as well, so I can re-request review later? (One line of comment sufficies.)

Reply:
Hmm right, but you have no way to know which library ERR_LIB number refers to,
therefore I did not see what the number is good for, except tell you that a ERR_LIB_USER
caused the error.
Only the builtin numbers < 128 have fixed meaning, everything >= 128 just means,
you can only guess which module caused it, because the string table is missing.
And moreover, if ERR_get_next_error_library() is called more than 128 times, the
assigned ERR_LIB for those silently collide with the built-in ERR_LIB_XX defines.

Reply:
The library/provider could compare the number with the one it registered to distinguish its own error by the ones from other libraries/providers. 🤔

Reply:
> And moreover, if ERR_get_next_error_library() is called more than 128 times, the
> assigned ERR_LIB for those silently collide with the built-in ERR_LIB_XX defines.

This is certainly an error which should be catched (in another PR).

Reply:
Yes, but the only problem I see, is that no one of the current users of
ERR_get_next_error_library() has any idea how to do a proper error handling,
when this function returns 0.

Reply:
> The library/provider could compare the number with the one it registered to distinguish its own errors by the ones from other libraries/providers.

Yes, in theory, but since the uniqueness is not guaranteed, anything that depends on
the uniqueness is pretty fragile.

Reply:
It's true that it's hard to find out what the error library number for an engine or provider is hard to get.  Than I'd say that's a bug / missing feature in itself (it shouldn't be that hard to add an accessor function for it), *not* a reason to obliterate the library numbers even more.

Reply:
> @levitte may I ask you that you add this as a formal review change request to #17975 as well, so I can re-request review later? (One line of comment sufficies.)

Done

Reply:
The 128 limit has been around for years, and has gone unnoticed as far as I know.  This is probably because people load only so many engines / providers at the same time (or have only so many OpenSSL compatible "libraries" in their application, which is also a possibility)...

Still, it's a valid question what should happen when that limit is reached!
I'd say that this is worth raising a separate issue.

Reply:
Hmm, in crypto/provider_core.c the function ERR_get_next_error_library();
is invoked here
https://github.com/openssl/openssl/blob/e4cdcb8bc44250aa4e0893dc4a7d64668f0fb949/crypto/provider_core.c#L1236
but it's result is never used here
https://github.com/openssl/openssl/blob/e4cdcb8bc44250aa4e0893dc4a7d64668f0fb949/crypto/provider_core.c#L1906-L1910
because ERR_GET_LIB(reason) is never zero, but instead most of the time ERR_LIB_PROV
but the reason code may be from a different openssl version.